README
=========================================================
Geant4 - an Object-Oriented Toolkit for Simulation in HEP
=========================================================

		  			ABALONE photosensor
		  			-------------------

----------------------------------------------------------------------------------------------------------------------------------------------------------------

--Building the simulation from source code--

% mkdir build
% cd build
% sudo cmake -DGeant4_DIR=<Your Geant4 install Directory (like /usr/share/geant4/geant4.10.07-install/lib/Geant4-10.7.1/)> <Your Directory>ABALONE_GEANT4/GEANT4
% sudo make install -j16 && sudo make

----------------------------------------------------------------------------------------------------------------------------------------------------------------

--Before running the simulation--

Change the save directory of the output files in the code in both ABEventAction.cc and ABEventAction.hh

Import the electric fields
% mkdir E_fields
% cp <directory of the electric fields from COMSOL>

and change the directory of the electric field map in the file ABElectricField.cc

Below the folder structrue for reference:
-> <My Directory>
----> ABALONE_GEANT4
----> GEANT4
----> E_field
----> results
-------- SiPM
-------- tracking

----------------------------------------------------------------------------------------------------------------------------------------------------------------

--Execute the simulation--

In 'interactive mode' with visualization:
% ./ABsimulation

In 'batch' mode from macro files:
% ./ABsimulation -m full_run.mac




ABALONE_GEANT4/History
-------------------------------------------------------------------

     =========================================================
     Geant4 - an Object-Oriented Toolkit for Simulation in HEP
     =========================================================

                    AB simulation History file
                    -----------------------
This file should be used by the G4 example coordinator to briefly
summarize all major modifications introduced in the code and keep
track of all tags.

     ----------------------------------------------------------
     * Reverse chronological order (last date on top), please *
     ----------------------------------------------------------
     
26/09/16 I. Hrivnacova (exampleB1-V10-02-01)
- Updated for renaming G4Parameter in G4Accumulable

20/09/16 J. Allison (exampleB1-V10-02-00)
- vis.mac: Added this to make "Envelope" transparent blue:
  # "Envelope" is transparent blue to represent water
  /vis/geometry/set/colour Envelope 0 0 0 1 .3

02/11/15 I. Hrivnacova (exampleB1-V10-01-03)
- B1Run class replaced with a code based on G4Parameter

04/05/15 I. Hrivnacova (exampleB1-V10-01-02)
- Coding guidelines: removed empty lines
     
23/04/15 mma (exampleB1-V10-01-01)
- RunAction : come back to previous formula
     
21/04/15 mma (exampleB1-V10-01-00)
- RunAction : correct calculation of rmsEdep
  
29/11/14 I. Hrivnacova
- Use G4endl instead of \n in G4cout;
  this makes each new line in the output on threads preceded with
  G4WTN >

06/11/14 I. Hrivnacova
- Removed G4UI_USE/G4VIS_USE tests and init.mac macro (not needed)
- Moved G4UIExecutive at the beginning of main() in all examples
- Perform run initialization in macros instead of main()

28/10/13 I. Hrivnacova (exampleB1-V09-06-06)
- Removed SetNumberOfThreads(4) from main (use Geant4 default)
     
26/10/13 mma (exampleB1-V09-06-05)
- Use /run/printProgress. Cleanup in EventAction

08/10/13 I. Hrivnacova (exampleB1-V09-06-04)
- Removed B1EventInformation for keeping maximum simplicity
- Improved documentation (added paragraph on Run::Merge())
- Code clean-up
     
09/06/13 I. Hrivnacova (exampleB1-V09-06-03)
- clarify local names in user actions
     
05/06/13 mma (exampleB1-V09-06-02)
- add section about ACTION INITALIZATION to README and .README
- update section DETECTOR RESPONSE

05/05/13 I. Hrivnacova  (exampleB1-V09-06-01)
- Migration for MT (by Makoto):
  Added B1ActionInitialization, B1EventInformation and B1Run classes
  and updated actions classes accordingly.
  README files still need to be updated.

15/01/13 I. Hrivnacova  (exampleB1-V09-06-00)
- Tag for a test only (g4svn update with svn 1.7.x)

13/11/12 I. Hrivnacova  (exampleB1-V09-05-03)
- Use QBBC physics list instead of QGSP_BIC_EMY, which becomes
  obsolete

02/11/12 J. Allison  (exampleB1-V09-05-01 and 02)
- README: Improved.

02/11/12 J. Allison  (exampleB1-V09-05-00)
- vis.mac: Improved view and added text, scale, logo etc. to scene.

14/11/11 I. Hrivnacova
- The first tagged version of the new B1 example
  (tagged in basic) 

31/10/11 I. Hrivnacova
- change volume names
- scoring in 1 volume only, with new scheme     
     
05/09/11 M. Maire, P. Gueye
- Created.


GEANT4/ABsimulation.cc
//
// ********************************************************************
// * License and Disclaimer                                           *
// *                                                                  *
// * The  Geant4 software  is  copyright of the Copyright Holders  of *
// * the Geant4 Collaboration.  It is provided  under  the terms  and *
// * conditions of the Geant4 Software License,  included in the file *
// * LICENSE and available at  http://cern.ch/geant4/license .  These *
// * include a list of copyright holders.                             *
// *                                                                  *
// * Neither the authors of this software system, nor their employing *
// * institutes,nor the agencies providing financial support for this *
// * work  make  any representation or  warranty, express or implied, *
// * regarding  this  software system or assume any liability for its *
// * use.  Please see the license in the file  LICENSE  and URL above *
// * for the full disclaimer and the limitation of liability.         *
// *                                                                  *
// * This  code  implementation is the result of  the  scientific and *
// * technical work of the GEANT4 collaboration.                      *
// * By using,  copying,  modifying or  distributing the software (or *
// * any work based  on the software)  you  agree  to acknowledge its *
// * use  in  resulting  scientific  publications,  and indicate your *
// * acceptance of all terms of the Geant4 Software license.          *
// ********************************************************************
//
//
/// \file ABsimulation.cc
/// \brief Main program of the AB simulation

#include "PhysicsList.hh"
#include "ABDetectorConstruction.hh"
#include "ABActionInitialization.hh"
#include "ABSteppingAction.hh"

#ifdef G4MULTITHREADED
#include "G4MTRunManager.hh"
#else
#include "G4RunManager.hh"
#endif

#include "time.h"
#include <stdlib.h>
#include <iostream>

#include "QGSP_BIC.hh"
#include "FTFP_BERT.hh"
#include "G4UImanager.hh"
#include "LBE.hh"
#include "QGSP_BIC.hh"
#include "G4StepLimiterPhysics.hh"
#include "G4EmPenelopePhysics.hh"
#include "G4OpticalPhysics.hh"
#include "G4EmStandardPhysics_option4.hh"
#include "G4RadioactiveDecayPhysics.hh"
#include "G4DecayPhysics.hh"
#include "ABStackingAction.hh"

#include "G4EmLivermorePhysics.hh"
#include "G4PhysListFactory.hh"

#include "G4VisExecutive.hh"
#include "G4UIExecutive.hh"

#include "Randomize.hh"

std::vector<int> photon_sipm_hit;

namespace {
	void PrintUsage() {
		//G4cerr << " Usage: " << G4endl;
		//G4cerr << " ABsimulation [-m macro ] [-u UIsession] [-t nThreads]" << G4endl;
		//G4cerr << "   note: -t option is available only for multi-threaded mode."
			//<< G4endl;
	}
}

//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......

int main(int argc, char** argv){

	if (argc > 7) {
		PrintUsage();
		return 1;
	}

	std::cerr << "     ========================================================== " << std::endl;
	std::cerr << "     ======= A Geant4 Monte Carlo Simulation on ABALONE ======= " << std::endl;
	std::cerr << "     ========================================================== \n" << std::endl;
	
	std::cerr << "                           %%%%%%%%%%%                   /         " << std::endl; //18
	std::cerr << "                       %%%%%%%%%%%%%%%%%%%%             /         " << std::endl; //17
	std::cerr << "                     %%%%%%%%%%%%%%%%%%%%%%%%          /          " << std::endl; //16
	std::cerr << "                    %%%%%%%%%%%%%%%%%%%%%%%%%%        /           " << std::endl; //15
	std::cerr << "                   %%%%%%%%%%%%%%%%%%%%%%%%%%%%      *            " << std::endl; //14
	std::cerr << "                  %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%                  " << std::endl; //13
	std::cerr << "                 %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%                 " << std::endl; //12
	std::cerr << "                %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%                " << std::endl; //11
	std::cerr << "               %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%               " << std::endl; //10
	std::cerr << "               %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%               " << std::endl; //9
	std::cerr << "              %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%              " << std::endl; //8
	std::cerr << "              %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%              " << std::endl; //7
	std::cerr << "              %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%              " << std::endl; //6
	std::cerr << "              %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%              " << std::endl; //5
	std::cerr << "             %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%             " << std::endl; //4
	std::cerr << "             %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%             " << std::endl; //3
	std::cerr << "             %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%             " << std::endl; //2
	std::cerr << "             %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%             " << std::endl; //1
	std::cerr << "             %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%             " << std::endl; //-1
	std::cerr << "             %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%             " << std::endl; //-2
	std::cerr << "             %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%             " << std::endl; //-3
	std::cerr << "             %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%             " << std::endl; //-4
	std::cerr << " \n\n\n ----------- WELCOME TO ABALONE Simulation (ver 2.0.1)! ----------- \n" << std::endl;
	std::cerr << " ** Program Developed by: Sze Chun Yiu @ Stockholm University " << std::endl;
	std::cerr << " ** Electric Field Map Calculated via COMSOL multiphysics \n\n" << std::endl;
	std::cerr << " Reference:                            \n\n [1] Ferenc, Daniel, et al. ABALONETM Photosensors for the IceCube experiment. Nuclear Instruments and Methods in Physics Research Section A: Accelerators, Spectrometers, Detectors and Associated Equipment (2018). \n" << std::endl;
	std::cerr << " [2] Ferenc, Daniel, Andrew Chang, and Marija Šegedin Ferenc. The Novel ABALONE Photosensor Technology : 4 - Year Long Tests of Vacuum Integrity, Internal Pumpingand Afterpulsing.arXiv preprint arXiv:1703.04546 (2017). \n\n" << std::endl;
	

	G4String macro;
	G4String session;

#ifdef G4MULTITHREADED
    
	G4int nThreads = 4;
#endif

    for (G4int i = 1; i < argc; i = i + 2) {
		if (G4String(argv[i]) == "-m") macro = argv[i+ 1];
		else if (G4String(argv[i]) == "-u") session = argv[i + 1];

#ifdef G4MULTITHREADED
		else if (G4String(argv[i]) == "-t") {
			nThreads = G4UIcommand::ConvertToInt(argv[i + 1]);
		}
#endif
		else {
			PrintUsage();
			return 1;
		}
	}

	// Detect interactive mode (if no macro provided) and define UI session
	//
	G4UIExecutive* ui=0;
	if (!macro.size()){
		ui=new G4UIExecutive(argc, argv, session);
	}

#ifdef G4MULTITHREADED
	G4MTRunManager* runManager = new G4MTRunManager;
	if (nThreads > 0) {
		runManager->SetNumberOfThreads(nThreads);
	}
#else
	G4RunManager* runManager=new G4RunManager;
#endif

	// Set mandatory initialization classes
	//
	std::cerr << "Initializing the Detector Consturction..." << std::endl;
	auto detConstruction = new ABDetectorConstruction("resources/sensl-microfc-60035-sm.properties","default");
	runManager->SetUserInitialization(detConstruction);

	std::cerr << "Initializing the Physics List..." << std::endl;

	//G4VUserPhysicsList* physics = new ABPhysicsList;
  	//runManager->SetUserInitialization(physics);
	runManager->SetUserInitialization(new PhysicsList);

//-----------------------------------------------------------------
	
	//G4VModularPhysicsList* physicsList=new QGSP_BIC;
	//physicsList->ReplacePhysics(new G4EmStandardPhysics_option4());

	//G4OpticalPhysics* opticalPhysics=new G4OpticalPhysics();
	//auto opticalParams=G4OpticalParameters::Instance();
	//opticalParams->SetWLSTimeProfile("delta");
	//opticalParams->SetScintYieldFactor(1.0);
	//opticalParams->SetScintExcitationRatio(0.0);
	//opticalParams->SetCerenkovMaxPhotonsPerStep(2000);
	//opticalParams->SetCerenkovMaxBetaChange(100.0);
	//opticalParams->SetCerenkovTrackSecondariesFirst(true);
	//opticalParams->SetScintTrackSecondariesFirst(true);


	//physicsList->RegisterPhysics( new G4RadioactiveDecay());
	//physicsList->RegisterPhysics(opticalPhysics);
	

	physicsList->RegisterPhysics(new G4RadioactiveDecayPhysics());
	physicsList->RegisterPhysics(new G4DecayPhysics());

	//G4RadioactiveDecay* theRadioactiveDecay=new G4RadioactiveDecay();
	//G4ProcessManager* pmanager = pmanager ->AddProcess(theRadioactiveDecay);

	
	//physicsList->RegisterPhysics(new G4RadioactiveDecayPhysics());


	//G4RadioactiveDecay* RadioactiveDecay=new G4RadioactiveDecay();
	//auto RadioactiveDecayParams=G4OpticalParameters::Instance();
	//RadioactiveDecayParams->SetAnalogueMonteCarlo(false);

  	//physicsList->RegisterPhysics(new G4RadioactiveDecayPhysics());

	//runManager->SetUserInitialization(physicsList);

//------------------------------------------------------------------------

	std::cerr << "Initializing the Actions..." << std::endl;
	auto actionInitialization=new ABActionInitialization();
	runManager->SetUserInitialization(actionInitialization);


	runManager->Initialize();

	auto UImanager=G4UImanager::GetUIpointer();
	UImanager->ApplyCommand("/process/verbose 0");
	UImanager->ApplyCommand("/vis/verbose errors");
	UImanager->ApplyCommand("/gps/verbose 0");
	UImanager->ApplyCommand("/control/verbose 0");


	CLHEP::HepRandom::setTheEngine(new CLHEP::RanecuEngine());
	time_t systime = time(NULL);
	long seed = (long)systime;
	CLHEP::HepRandom::setTheSeed(seed);

	CLHEP::HepRandomEngine* theEngine = CLHEP::HepRandom::getTheEngine();
	auto theSeed = theEngine->getSeed();
	G4cout << "Random seed check: " << theSeed << G4endl;


	// Initialize visualization
	std::cerr << "Initializing Visualization..." << std::endl;
	auto visManager = new G4VisExecutive;
	visManager->Initialize();

	// Get the pointer to the User Interface manager
	std::cerr << "Initializing User Interface Manager..." << std::endl;
	
	

	// Process macro or start UI session

	if (macro.size()) { //macro.size()
		// batch mode
		G4String command = "/control/execute ";
		G4String mac_file = macro;
		UImanager->ApplyCommand(command + mac_file);
	}
	else {
		// interactive mode : define UI session
		UImanager->ApplyCommand("/control/execute vis.mac");
		if (ui->IsGUI()) {
			UImanager->ApplyCommand("/control/execute gui.mac");
		}
		ui->SessionStart();
		delete ui;
	}

	delete visManager;
	delete runManager;
}

//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo.....
